crates/sui-core/src/authority.rs

```diff
@@ -4649,6 +4651,112 @@ impl AuthorityState {
+        if epoch_store.epoch() != epoch {
+            error!(
+                "BUG: dropping randomness for epoch {epoch}, round {round}, because we are in epoch {}",
+                epoch_store.epoch()
+            );
+            return;
+        }
+        let transaction = VerifiedTransaction::new_randomness_state_update(
+            epoch,
+            round,
+            bytes,
+            epoch_store
+                .epoch_start_config()
+                .randomness_obj_initial_shared_version()
+                .expect("randomness state obj must exist"),
+        );
+        debug!(
+            "created randomness state update transaction with digest: {:?}",
+            transaction.digest()
+        );
+        let transaction = VerifiedExecutableTransaction::new_system(transaction, epoch);
+        let key = transaction.key();
+
+        // RandomnessStateUpdates are not sent through consensus, so we have to add the user
+        // signatures here once the digest becomes known.
+        let Ok(tables) = epoch_store.tables() else {
+            debug!("dropping randomness because current epoch is ending");
+            return;
+        };
+        let mut batch = tables.pending_execution.batch();
+        if let Err(e) =
+            epoch_store.finish_consensus_certificate_process_with_batch(&mut batch, &transaction)
+        {
+            warn!("dropping randomness: {e:?}");
+            return;
+        }
+        batch.write().expect("storage should not fail");
+
+        // Send transaction to TransactionManager for execution.
+        if let Err(e) = self
+            .authority_state
+            .transaction_manager()
+            .enqueue(vec![transaction], &epoch_store)
+        {
+            error!("BUG: failed to enqueue randomness state update transaction: {e:?}",);
+        }
+
+        let authority_state = self.authority_state.clone();
+        spawn_monitored_task!(async move {
+            // Wait for transaction execution in a separate task, to avoid deadlock in case of
+            // out-of-order randomness generation. (Each RandomnessStateUpdate depends on the
+            // output of the RandomnessStateUpdate from the previous round.)
+            let Ok(mut effects) = authority_state
+                .execution_cache
+                .notify_read_executed_effects(&[key])
+                .await
+            else {
+                panic!("failed to get effects for randomness state update transaction at epoch {epoch}, round {round}");
+            };
+            let effects = effects.pop().expect("should return effects");
+            if *effects.status() != ExecutionStatus::Success {
+                panic!("failed to execute randomness state update transaction at epoch {epoch}, round {round}: {effects:?}");
+            }
+            debug!("successfully executed randomness state update transaction at epoch {epoch}, round {round}");
```

Comment:
this is a lot of code just to write out a log. If we log the tx digest before enqueuing it, we will be able to find all the logs for that transaction, so i'm not sure this adds much utility

Reply:
It doesn't just write a log, it panics on failure

Reply:
then it would be much simpler to just panic in `setup_randomness_state_update` in `execution_engine.rs`. btw, we don't panic on failure for any other of the system transactions. Perhaps we should. @lxfind

Reply:
What happens if we don't panic here?

Reply:
The randomness state update won't happen and then we'd have multiple commits' worth of tx reading from the same randomness version, which I believe is a security problem (eg if a validator found a bug to make these update tx fail, and then could deliberately trigger it so that they could then predict the randomness used by future user tx)

Reply:
@mystenmark is that function (`programmable_transactions::execution::execute` called by `setup_randomness_state_update`) guaranteed to be the only place a tx can fail to execute?